src/auth/src/credentials/internal/sts_exchange.rs

```diff
@@ -0,0 +1,442 @@
+            client_id: None,
+            client_secret: None,
+        }
+    }
+}
+
+impl ClientAuthentication {
+    // Add authentication to a Secure Token Service exchange request.
+    // Modifies either the passed headers or form parameters
+    // depending on the desired authentication format.
+    #[allow(dead_code)]
+    pub fn inject_auth(&self, headers: &mut http::HeaderMap, params: &mut HashMap<&str, String>) {
+        if let (Some(client_id), Some(client_secret)) =
+            (self.client_id.clone(), self.client_secret.clone())
+        {
+            match self.auth_style {
+                ClientAuthStyle::InHeader => {
+                    let plain_header = format!("{}:{}", client_id, client_secret);
+                    let encoded =
+                        base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(plain_header);
+                    let header = http::HeaderValue::from_str(format!("Basic {encoded}").as_str());
+                    if let Ok(value) = header {
+                        headers.insert("Authorization", value);
+                    }
+                }
+                _ => {
+                    params.insert("client_id", client_id);
+                    params.insert("client_secret", client_secret);
+                }
+            }
+        }
+    }
+}
+
+/// Information required to perform an oauth2 token exchange with the provided endpoint.
+#[allow(dead_code)]
+#[derive(Default)]
+pub struct ExchangeTokenRequest {
+    pub url: String,
+    pub authentication: ClientAuthentication,
+    pub headers: http::HeaderMap,
+    pub resource: Option<String>,
+    pub subject_token: String,
+    pub subject_token_type: String,
+    pub audience: Option<String>,
+    pub scope: Vec<String>,
+    pub actor_token: Option<String>,
+    pub actor_token_type: Option<String>,
+    pub extra_options: Option<HashMap<String, String>>,
+}
+
+/// Information required to perform the token exchange using a refresh token flow.
+#[allow(dead_code)]
+#[derive(Default)]
+pub struct RefreshAccessTokenRequest {
+    pub url: String,
+    pub authentication: ClientAuthentication,
+    pub headers: http::HeaderMap,
+    pub refresh_token: String,
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use httptest::{Expectation, Server, matchers::*, responders::*};
+    use tokio_test::assert_err;
+    type TestResult = std::result::Result<(), Box<dyn std::error::Error>>;
+
+    #[tokio::test]
+    async fn exchange_token() -> TestResult {
+        let authentication = ClientAuthentication {
+            auth_style: ClientAuthStyle::InHeader,
+            client_id: Some("client_id".to_string()),
+            client_secret: Some("supersecret".to_string()),
+        };
+        let response_body = r#"{"access_token":"an_example_token","issued_token_type":"urn:ietf:params:oauth:token-type:access_token","token_type":"Bearer","expires_in":3600,"scope":"https://www.googleapis.com/auth/cloud-platform"}"#;
+
```

Comment:
Consider converting the body to a `serde_json::Value` and then comparing those. That will be more robust as fields get reordered.

Reply:
Oh, I see, this is an input to the test. Nevertheless, I would write something like this for readability:

```suggestion
        let response_body = json!({
            "access_token":"an_example_token",
            "issued_token_type":"urn:ietf:params:oauth:token-type:access_token",
            "token_type":"Bearer",
            "expires_in":3600,
            "scope":"https://www.googleapis.com/auth/cloud-platform"
        }).to_string();
```